src/lib.rs

```diff
@@ -1006,6 +1016,21 @@ impl<'a, 'b> EventCtx<'a, 'b> {
     pub fn request_focus(&mut self) {
         self.base_state.request_focus = true;
     }
+
+    /// Request an animation frame.
+    pub fn request_anim_frame(&mut self) {
+        self.base_state.request_anim = true;
+    }
+
+    /// Request a timer event.
+    ///
+    /// The return value is an id, which can be used to associate the
+    /// request with the event. The id is guaranteed not to be 0, and can
+    /// also be assumed to be a small integer.
+    pub fn request_timer(&mut self, deadline: Instant) -> usize {
```

Comment:
I might call this `add_timer`. 'request' is tricky; can requests be refused? Are we given a timer if our request is accepted? etc.

There's also the question of what the argument should be. There's a reasonable case that this could just be a duration, which is taken relative to the time when the call is made. This would also allow for things like repeating timers, and prevent things like scheduling timers for times in the past.

If we make our rumoured changes to `Action`, we can also associate an `Action` with a timer which would be a nicer API.

We might also want to consider giving timers a name, which would let callers use const strs to identify timers, instead of needing to stash the returned token.

Reply:
I'm fine with changing the name to `add_timer`. The existing name is designed to be parallel to `request_anim_frame`, which basically comes from web world.

My understanding is that instant is the gold standard for timers. Fuchsia iterated on this a bit. One motivation is that if you want to blink something at a fixed rate, you repeatedly add the period to the base instant. This is especially important if the underlying mechanism is sloppy. (As an aside, I'm personally interested in precise time for, eg, music production, but this is an extremely difficult problem, so it needs to go on the back burner)

I think names are *highly* problematic, as there's no mechanism to prevent collisions. Remember, widgets don't have identity, though this is perhaps a question we could come back to.

I'm open to making `Action` include timers, but confess to being very fuzzy what that would actually look like. Maybe more clarity will come when we wire up other things such as context menus.

So, in sum, while this API is perhaps not the absolute most friendly (the user needs to stash the id and filter on it), I'll defend it as being a pretty solid way to get things done, requiring a smidge of extra work on the part of the widget author that I don't see any easy way to remove without adding significant extra complexity in the system.

Reply:
Not sure I understand, about `Instant`: if you want to blink something at a fixed rate with `Duration`, you just set a repeating timer, and the framework reschedules whenever it delivers; and on Cocoa, at least, we can just schedule a repeating timer directly. (Cocoa provides API for using both duration (the default) and also firing at a specific datetime).

named timers would require some convention like reverse dns, for namespacing. I'm happy not to do this, but I'm not sure it's a huge problem in practice; any timer names used by the framework would be collision-resilient, and it would be up to the application developer to pick their own names appropriately.

Reply:
Having an explicit repeating timer is one way to do this, and very common in timer API's. I believe it complicates the API for not too much benefit - you then need to be able to cancel the timer. But this might just be my preference for a small number of simple primitives.

I think for our use cases, duration and instant aren't that much different. I think for advanced uses, instant is more robust.

Reply:
I would suggest that the prevalence of API for repeated timers suggests they may be useful.

I also think that in 99% of cases, this API will be called with some duration relative to `Instant::now()`, that is, the `Instant` bit is redundant.

That said, it's really a wash unless we want repeating timers; the only advantage to `Duration` is requiring one less import and one less line of boilerplate; no big deal, but I do think these things are worth weighing.